src/main/kotlin/no/nav/tilleggsstonader/sak/vilkår/stønadsperiode/StønadsperiodeService.kt

```diff
@@ -1,38 +1,90 @@
+import org.springframework.transaction.annotation.Transactional
 import java.util.UUID
 
 @Service
 class StønadsperiodeService(
+    private val behandlingService: BehandlingService,
     private val stønadsperiodeRepository: StønadsperiodeRepository,
     private val vilkårperiodeService: VilkårperiodeService,
 ) {
     fun hentStønadsperioder(behandlingId: UUID): List<StønadsperiodeDto> {
         return stønadsperiodeRepository.findAllByBehandlingId(behandlingId).tilSortertDto()
     }
 
+    @Transactional
     fun lagreStønadsperioder(behandlingId: UUID, stønadsperioder: List<StønadsperiodeDto>): List<StønadsperiodeDto> {
-        // TODO valider behandling er ikke låst
+        feilHvis(behandlingService.hentBehandling(behandlingId).status.behandlingErLåstForVidereRedigering()) {
+            "Kan ikke lagre stønadsperioder når behandlingen er låst"
+        }
         validerStønadsperioder(behandlingId, stønadsperioder)
-        stønadsperiodeRepository.deleteAll() // TODO skal vi finne og oppdatere de som har blitt oppdatert/slettet
-        return stønadsperiodeRepository.insertAll(
-            stønadsperioder.map {
-                Stønadsperiode(
-                    id = it.id ?: UUID.randomUUID(),
-                    behandlingId = behandlingId,
+
+        val tidligereStønadsperioder = stønadsperiodeRepository.findAllByBehandlingId(behandlingId)
+
+        slettStønadsperioder(tidligereStønadsperioder, stønadsperioder)
+        val oppdaterteStønadsperioder = oppdaterStønadsperioder(tidligereStønadsperioder, stønadsperioder)
+        val nyeStønadsperioder = nyeStønadsperioder(behandlingId, stønadsperioder, tidligereStønadsperioder)
+        return (nyeStønadsperioder + oppdaterteStønadsperioder).tilSortertDto()
+    }
+
+    private fun slettStønadsperioder(
+        tidligereStønadsperioder: List<Stønadsperiode>,
+        stønadsperioder: List<StønadsperiodeDto>,
+    ) {
+        val gjenståendePerioder = stønadsperioder.mapNotNull { it.id }.toSet()
+        val perioderTilSletting = tidligereStønadsperioder.filterNot { gjenståendePerioder.contains(it.id) }
+        stønadsperiodeRepository.deleteAllById(perioderTilSletting.map { it.id })
+    }
+
+    private fun oppdaterStønadsperioder(
+        tidligereStønadsperioder: List<Stønadsperiode>,
+        stønadsperioder: List<StønadsperiodeDto>,
+    ): List<Stønadsperiode> {
+        val stønadsperioderPåId =
+            stønadsperioder.mapNotNull { periode -> periode.id?.let { id -> id to periode } }.toMap()
+        val perioderTilOppdatering = tidligereStønadsperioder.mapNotNull { tidligereStønadsperiode ->
+            stønadsperioderPåId[tidligereStønadsperiode.id]?.let {
+                tidligereStønadsperiode.copy(
                     fom = it.fom,
                     tom = it.tom,
                     målgruppe = it.målgruppe,
                     aktivitet = it.aktivitet,
                 )
-            },
-        ).tilSortertDto()
+            }
```

Comment:
Kanskje unødvendig, for vil jo ikke skje med denne koden, men man kunne kanskje slengt på ` ?: error("...")`

Reply:
Har med `mapNotNull` og oppdaterer kun perioder som finnes fra før